### PR TITLE
Support OpenMP get_thread_num

### LINKED ISSUE #117: OpenMP - Check for paths guarded by omp_get_thread_num
**Is your feature request related to a problem? Please describe.**
Currently the tool will report false positives for cases where a race is prevented by a branch condition that checks for a specific thread id.

```cpp
int shared = 0;

#pragma omp parallel
{
  if (omp_get_thread_num() == 1) {
    shared++; // No race only thread 1 will execute this line
  }
}
```
We do not recognize that the condition ensures only one thread will execute this branch.

**Describe the solution you'd like**
Add an omp specific analysis that identifies branch conditions for racing events and checks for the following conditional pattern:

```cpp
omp_get_thread_num() == SomeConstant
```
If both accesses are guarded by the same constant, do not report the race.

**Additional context**
Doing this at a full scale might be difficult. If it is too difficult, we can only implement the check withing a single function and create a new issue to make the check work for the whole stack trace.

See DRB cases 51, 75, and 171 for more examples.

### PR DESCRIPTION
**Pull request by BradSwain**
Closes #117 

This should handle the most common usages, except inter-procedural cases. I created a new issue (#208) for handling inter-procedural cases.

For some reason DRB171 crashes in the same place as DRB058. I added a comment and updated the issue opened for the DRB058 crash.

### CODE REVIEW

File: src/Analysis/OpenMPAnalysis.cpp [hunk 1]
```diff
@@ -697,6 +699,150 @@ bool OpenMPAnalysis::inSameReduce(const Event *event1, const Event *event2) cons
...
+  }
+
+  return result;
+}
+
+// Get list of blocks guarded by the true case of this branch.
+// Start by assuming the true block is guarded
+// Iterate from the true block until we find a block that has an unguarded predecessor
+// Cannot handle loops
+std::optional<std::set<const llvm::BasicBlock *>> getGuardedBlocks(const llvm::BranchInst *branchInst) {
```
Comment (addison-coderrect) [inline src/Analysis/OpenMPAnalysis.cpp:737]: std::optional needed here?

Comment (BradSwain) [reply src/Analysis/OpenMPAnalysis.cpp:737]: good catch, this was left behind from on older implementation that tried to handle inter-procedural as well. I will update.

**Comment by BradSwain** [comment]
@addison-coderrect if everything looks good feel free to merge.

**Review by addison-coderrect — APPROVED** [review]
LGTM

**Pull request merged**